tico/experimental/quantization/algorithm/gptq/quantizer.py

```diff
@@ -50,176 +68,201 @@ def prepare(
+                    layer(*cache_args_batch, **cache_kwargs_batch)
+
+                # Remove handles
                 for h in handles:
                     h.remove()
-                # Quantize each submodule using the collected calibration data.
+
+                # 3) Quantize each submodule
                 for name in subset:
                     if gptq_conf.verbose:
-                        print(l_idx, name)
-                        print("Quantizing ...")
+                        print(f"[Layer {l_idx}] {name} -> Quantizing ...")
                     gptq[name].fasterquant(
                         percdamp=0.01,
                         groupsize=-1,
                         actorder=True,
                         static_groups=False,
                         verbose=gptq_conf.verbose,
                     )
-                    quantizers["model.layers.%d.%s" % (l_idx, name)] = gptq[
-                        name
-                    ].quantizer
+                    quantizers[f"model.layers.{l_idx}.{name}"] = gptq[name].quantizer
                     gptq[name].free()
-            """
-            Execute the quantized layer with the calibration inputs to obtain ouptuts
-             that will serve as inputs for the next layer.
 
-            This ensures that the quantization effects are correctly propagated to subsequent
-             layers.
-            """
-            batch_num = self.cache_kwargs.pop("batch_num")
+            # 4) After quantization, re-run the layer to produce outputs for the next layer
             for batch_idx in range(batch_num):
                 cache_args_batch = gather_single_batch_from_list(
                     self.cache_args, batch_idx
                 )
                 cache_kwargs_batch = gather_single_batch_from_dict(
                     self.cache_kwargs, batch_idx
                 )
-                outs = layer(*cache_args_batch, **cache_kwargs_batch)[0]
+                outs = layer(*cache_args_batch, **cache_kwargs_batch)
+                # LLaMA's decoder layer return type differs across Transformers versions:
+                # some return a tuple (hidden_states, ...), others return just a tensor.
+                # This line ensures we always take the first element when it's a tuple.
+                outs = outs[0] if isinstance(outs, tuple) else outs
                 # Update inputs for next iteration.
                 self.cache_args[0][batch_idx] = outs
-            self.cache_kwargs["batch_num"] = batch_num
 
             if torch.cuda.is_available():
                 torch.cuda.empty_cache()
+
         # Restore the original cache configuration.
-        if hasattr(model, "config"):
-            model.config.use_cache = use_cache
+        if orig_use_cache is not None:
+            model.config.use_cache = orig_use_cache
+
+        # Clear caches to free memory
+        self.cache_args.clear()
+        self.cache_kwargs.clear()
+        self.num_batches = 0
```

Comment:
I assume that after `convert`, `prepare` can be called again by cleaning the context with these lines.
Then, how about wrapping these as an initialize function and use here or at the beginning of `prepare`?